syntax = "proto3";

package order_executor;

service OrderExecutorService {
  // Called by a lower ID replica to start an election
  rpc StartElection(ElectionRequest) returns (ElectionResponse);

  // Called to announce who won the election
  rpc AnnounceLeader(LeaderAnnouncement) returns (Ack);

  // Called by the leader to dequeue and process an order
  rpc DequeueOrder(OrderRequest) returns (OrderResponse);
}

message ElectionRequest {
  int32 sender_id = 1;
}

message ElectionResponse {
  bool acknowledged = 1;
}

message LeaderAnnouncement {
  int32 leader_id = 1;
}

message Ack {
  bool received = 1;
}

message OrderRequest {
  string dummy = 1; // optional placeholder
}

message OrderResponse {
  string message = 1;
}
